### PR TITLE
Fix SFP Support and add Platform as PE feature

### PR DESCRIPTION
**Pull request by jahofmann**
This pull request adds the following new platform:
 - AU50: HBM attached in global address mode.

This pull request fixes/adds SFP+ for the following platforms:
 - VCU118
 - NetFPGA SUME: New tool to program the clock generator via I2C included.
 - AU50

These changes rely on a fix for #278 which is not included.

Finally, the rust runtime now supports treating platform components as PEs (Commit 03d717e). This way it is easy to add e.g. clock programmers to platforms and control them with the convenient PE interface. Furthermore, there is an unsafe function to provide direct access to the platform component memory which can be used for tools like tapasco-debug.

### CODE REVIEW

File: toolflow/vivado/common/common_2020.1.tcl [hunk 1]
```diff
@@ -27,3 +27,4 @@ dict set stdcomps   ultra_ps         vlnv   "xilinx.com:ip:zynq_ultra_ps_e:3.3"
 dict set stdcomps   xxv_ethernet     vlnv   "xilinx.com:ip:xxv_ethernet:3.2"
 dict set stdcomps   system_cache     vlnv   "xilinx.com:ip:system_cache:5.0"
 dict set stdcomps   axi_cache        vlnv   "xilinx.com:ip:system_cache:5.0"
+dict set stdcomps   xxv_ethernet     vlnv   "xilinx.com:ip:xxv_ethernet"
```
Comment (cahz) [inline toolflow/vivado/common/common_2020.1.tcl:30]: Duplicate to line 27?

File: toolflow/vivado/platform/AU50/AU50.tcl [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+#
+
+namespace eval platform {
+  set platform_dirname "AU50"
+  set pcie_width "x16"
+
+  if { [::tapasco::vivado_is_newer "2020.1"] == 0 } {
```
Comment (cahz) [inline toolflow/vivado/platform/AU50/AU50.tcl:24]: Is the U50 now directly available in Vivado? In branch `feature/alveo-u50` I had to use the XilinxBoardStore...

File: toolflow/vivado/platform/AU50/AU50.tcl [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+    puts "Creating HBM controllers ..."
+
+    set old_inst [current_bd_instance .]
+    set mig [create_bd_cell -type hier ${name}]
+    current_bd_instance $mig   
+
+
+    set hbm [tapasco::ip::create_hbm_controller ${name}_internal]
+
+set hbm_properties [list \
```
Comment (cahz) [inline toolflow/vivado/platform/AU50/AU50.tcl:58]: Please fix the indentation here

File: toolflow/vivado/platform/netfpga_sume/board.xdc [hunk 4]
```diff
@@ -20,6 +20,8 @@
 #The following two properties should be set for every design
 set_property CFGBVS GND [current_design]
 set_property CONFIG_VOLTAGE 1.8 [current_design]
+set_property BITSTREAM.GENERAL.COMPRESS TRUE [current_design]
+set_property BITSTREAM.CONFIG.UNUSEDPIN Pullup [current_design]
```
Comment (cahz) [inline toolflow/vivado/platform/netfpga_sume/board.xdc:24]: Is the pullup safe? What is the reason for it?

**Review by cahz — COMMENTED** [review]
Thanks for the contribution. I haven't looked at it in detail, but at first glance it looks good. I have a few comments and suggestions.

File: toolflow/vivado/platform/AU50/AU50.tcl (hunk 2 above)
Comment (jahofmann) [reply toolflow/vivado/platform/AU50/AU50.tcl:24]: Nope, that's part of the users Vivado installation. Not sure if adding it to TaPaSCo is a good solution to this problem. I could add an information where to get the necessary files.

File: toolflow/vivado/platform/netfpga_sume/board.xdc (hunk 4 above)
Comment (jahofmann) [reply toolflow/vivado/platform/netfpga_sume/board.xdc:24]: That comes from the reference XDC, so I hope it is safe... Had some stability issues with one of my boards and tried to rule out floating pins.

File: toolflow/vivado/platform/AU50/AU50.tcl (hunk 2 above)
Comment (cahz) [reply toolflow/vivado/platform/AU50/AU50.tcl:24]: Then maybe we can include my changes to automatically download it from the XilinxBoardStore. We should avoid adding the board files to the TaPaSCo repo.

File: toolflow/vivado/platform/netfpga_sume/board.xdc (hunk 4 above)
Comment (cahz) [reply toolflow/vivado/platform/netfpga_sume/board.xdc:24]: From the reference XDC seems good :+1:

File: toolflow/vivado/platform/AU50/AU50.tcl (hunk 2 above)
Comment (jahofmann) [reply toolflow/vivado/platform/AU50/AU50.tcl:24]: Are you sure the AU50 is still included by default? I can only find it for very specific Vivado versions, e.g. https://github.com/Xilinx/XilinxBoardStore/tree/2020.1.1/boards/Xilinx/au50/production/1.0 :/

Comment (cahz) [reply toolflow/vivado/platform/AU50/AU50.tcl:28]: You are right :-/ Back then I thought that it was this way because it is a new card. But apparently, Xilinx really does not want people to use the RTL flow...
The board files have Apache 2.0 license, which could be included in our repo.

Comment (wirthjohannes) [reply toolflow/vivado/platform/AU50/AU50.tcl:28]: The board files seem to be version specific. When using them with Vivado 2020.1 it works fine, but with Vivado 2020.2 it fails at the `apply_bd_automation` statements for the PCIe core.
So adding them to the repo is probably not an option

File: toolflow/vivado/platform/pcie/pcie_base.tcl [hunk 5]
```diff
@@ -17,358 +17,365 @@
...
+    CONFIG.USE_SAFE_CLOCK_STARTUP {false} \
+    CONFIG.CLKOUT1_REQUESTED_OUT_FREQ [tapasco::get_design_frequency] \
+    CONFIG.USE_LOCKED {true} \
+    CONFIG.USE_RESET {true} \
+    CONFIG.RESET_TYPE {ACTIVE_LOW} \
+    CONFIG.RESET_PORT {resetn} \
+    CONFIG.PRIM_SOURCE {No_buffer} \
+    ] $design_clk_wiz
+
+  connect_bd_net [get_bd_pins $design_clk_wiz/resetn] [get_bd_pins -regexp $mig/((init_calib_complete)|(c0_init_calib_complete))]
```
Comment (wirthjohannes) [inline toolflow/vivado/platform/pcie/pcie_base.tcl:263]: Why did you change this line? (also included `mmcm_locked` as possible ouput signal for the MIG before)
This breaks the new U50 because there the "MIG" has no `init_calib_complete` only the `mmcm_locked` output

Comment (jahofmann) [reply toolflow/vivado/platform/pcie/pcie_base.tcl:263]: I have to check what happened here. My U50.tcl creates an `mmcm_locked` output and I've been using it for quite some time successfully... Might have been some rebase issue with the master branch.

File: toolflow/vivado/platform/AU50/AU50.tcl (hunk 2 above)
Comment (jahofmann) [reply toolflow/vivado/platform/AU50/AU50.tcl:28]: 2020.1 was the only version I checked it with. This is version to version difference is very annoying :/

What kind of error are you seeing?

Comment (wirthjohannes) [reply toolflow/vivado/platform/AU50/AU50.tcl:28]: The first `apply_bd_automation` in `create_pcie_core` fails with the following:
```
ERROR: [IP_Flow 19-3461] Value 'GTH_Quad_128' is out of the range for parameter 'select quad(select_quad)' for BD Cell 'host/axi_pcie3_0' . Valid values are - GTY_Quad_227, GTY_Quad_226, GTY_Quad_225, GTY_Quad_224
ERROR: [BD 41-245] set_property error - Value 'GTH_Quad_128' is out of the range for parameter 'select quad(select_quad)' for BD Cell 'host/axi_pcie3_0' . Valid values are - GTY_Quad_227, GTY_Quad_226, GTY_Quad_225, GTY_Quad_224
Customization errors found on 'host/axi_pcie3_0'. Restoring to previous valid configuration.
```

Comment (jahofmann) [reply toolflow/vivado/platform/AU50/AU50.tcl:28]: Did they really change the Quad naming? What an odd thing to do.
